File: benchmark/nodejs/n_body.js
const PI = 3.141592653589793;
const SOLAR_MASS = 4 * PI * PI;
const DAYS_PER_YEAR = 365.24;

function advance(bodies, dt) {
    const n = bodies.length;
    for (let i = 0; i < n; i++) {
        const bi = bodies[i];
        for (let j = i + 1; j < n; j++) {
            const bj = bodies[j];
            const dx = bi[0] - bj[0];
            const dy = bi[1] - bj[1];
            const dz = bi[2] - bj[2];
            
            const dsq = dx*dx + dy*dy + dz*dz;
            const distance = Math.sqrt(dsq);
            const mag = dt / (dsq * distance);
            
            bi[3] -= dx * bj[6] * mag;
            bi[4] -= dy * bj[6] * mag;
            bi[5] -= dz * bj[6] * mag;
            
            bj[3] += dx * bi[6] * mag;
            bj[4] += dy * bi[6] * mag;
            bj[5] += dz * bi[6] * mag;
        }
    }
    for (let i = 0; i < n; i++) {
        const bi = bodies[i];
        bi[0] += dt * bi[3];
        bi[1] += dt * bi[4];
        bi[2] += dt * bi[5];
    }
}

function energy(bodies) {
    let e = 0.0;
    const n = bodies.length;
    for (let i = 0; i < n; i++) {
        const bi = bodies[i];
        e += 0.5 * bi[6] * (bi[3]*bi[3] + bi[4]*bi[4] + bi[5]*bi[5]);
        for (let j = i + 1; j < n; j++) {
            const bj = bodies[j];
            const dx = bi[0] - bj[0];
            const dy = bi[1] - bj[1];
            const dz = bi[2] - bj[2];
            const distance = Math.sqrt(dx*dx + dy*dy + dz*dz);
            e -= (bi[6] * bj[6]) / distance;
        }
    }
    return e;
}

const bodies = [
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, SOLAR_MASS],
    [4.84143144246472090, -1.16032004402742839, -0.103622044471123109,
     1.66007664274403694 * 1e-3 * DAYS_PER_YEAR, 7.69901118419740425 * 1e-3 * DAYS_PER_YEAR, -6.90460016972063023 * 1e-5 * DAYS_PER_YEAR,
     9.54791938424326609 * 1e-4 * SOLAR_MASS],
    [8.34336671824457987, 4.12479856412430479, -0.403523417114321381,
     -2.76742510726862411 * 1e-3 * DAYS_PER_YEAR, 4.99852801234917238 * 1e-3 * DAYS_PER_YEAR, 2.30417297573763929 * 1e-5 * DAYS_PER_YEAR,
     2.85885980666130812 * 1e-4 * SOLAR_MASS],
    [12.8943695621391310, -15.1111514016986312, -0.223307578892655734,
     2.96460137564761618 * 1e-3 * DAYS_PER_YEAR, 2.37847173959480950 * 1e-3 * DAYS_PER_YEAR, -2.96589568540237556 * 1e-5 * DAYS_PER_YEAR,
     4.36624404335156298 * 1e-5 * SOLAR_MASS],
    [15.3796971148509165, -25.9193146099879641, 0.179258772950371181,
     2.68067772490389322 * 1e-3 * DAYS_PER_YEAR, 1.62824170038242295 * 1e-3 * DAYS_PER_YEAR, -9.51592254519715870 * 1e-5 * DAYS_PER_YEAR,
     5.15138902046611451 * 1e-5 * SOLAR_MASS]
];

let px = 0.0, py = 0.0, pz = 0.0;
for (let i = 0; i < bodies.length; i++) {
    px += bodies[i][3] * bodies[i][6];
    py += bodies[i][4] * bodies[i][6];
    pz += bodies[i][5] * bodies[i][6];
}
bodies[0][3] = -px / SOLAR_MASS;
bodies[0][4] = -py / SOLAR_MASS;
bodies[0][5] = -pz / SOLAR_MASS;

const start = Date.now();
const e1 = energy(bodies);
for (let i = 0; i < 10000; i++) {
    advance(bodies, 0.01);
}
const e2 = energy(bodies);
const end = Date.now();

console.log(`Result: ${e1.toFixed(9)}_${e2.toFixed(9)}`);
console.log(`Time: ${end - start}ms`);

File: benchmarks/src/python/n_body.py
PI = 3.141592653589793
SOLAR_MASS = 4 * PI * PI
DAYS_PER_YEAR = 365.24

def advance(bodies, dt):
    for i in range(len(bodies)):
        body_i = bodies[i]
        for j in range(i + 1, len(bodies)):
            body_j = bodies[j]
            dx = body_i[0] - body_j[0]
            dy = body_i[1] - body_j[1]
            dz = body_i[2] - body_j[2]
            
            distance_sq = dx*dx + dy*dy + dz*dz
            mag = dt / (distance_sq * (distance_sq ** 0.5))
            
            body_i[3] -= dx * body_j[6] * mag
            body_i[4] -= dy * body_j[6] * mag
            body_i[5] -= dz * body_j[6] * mag
            
            body_j[3] += dx * body_i[6] * mag
            body_j[4] += dy * body_i[6] * mag
            body_j[5] += dz * body_i[6] * mag
            
    for body in bodies:
        body[0] += dt * body[3]
        body[1] += dt * body[4]
        body[2] += dt * body[5]

def energy(bodies):
    e = 0.0
    for i in range(len(bodies)):
        body_i = bodies[i]
        e += 0.5 * body_i[6] * (body_i[3]*body_i[3] + body_i[4]*body_i[4] + body_i[5]*body_i[5])
        for j in range(i + 1, len(bodies)):
            body_j = bodies[j]
            dx = body_i[0] - body_j[0]
            dy = body_i[1] - body_j[1]
            dz = body_i[2] - body_j[2]
            distance = (dx*dx + dy*dy + dz*dz) ** 0.5
            e -= (body_i[6] * body_j[6]) / distance
    return e

def main(n):
    # Format: [x, y, z, vx, vy, vz, mass]
    bodies = [
        # Sun
        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, SOLAR_MASS],
        # Jupiter
        [4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
         1.66007664274403694e-03 * DAYS_PER_YEAR, 7.69901118419740425e-03 * DAYS_PER_YEAR, -6.90460016972063023e-05 * DAYS_PER_YEAR,
         9.54791938424326609e-04 * SOLAR_MASS],
        # Saturn
        [8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
         -2.76742510726862411e-03 * DAYS_PER_YEAR, 4.99852801234917238e-03 * DAYS_PER_YEAR, 2.30417297573763929e-05 * DAYS_PER_YEAR,
         2.85885980666130812e-04 * SOLAR_MASS],
        # Uranus
        [1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
         2.96460137564761618e-03 * DAYS_PER_YEAR, 2.37847173959480950e-03 * DAYS_PER_YEAR, -2.96589568540237556e-05 * DAYS_PER_YEAR,
         4.36624404335156298e-05 * SOLAR_MASS],
        # Neptune
        [1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
         2.68067772490389322e-03 * DAYS_PER_YEAR, 1.62824170038242295e-03 * DAYS_PER_YEAR, -9.51592254519715870e-05 * DAYS_PER_YEAR,
         5.15138902046611451e-05 * SOLAR_MASS]
    ]

    px, py, pz = 0.0, 0.0, 0.0
    for b in bodies:
        px += b[3] * b[6]
        py += b[4] * b[6]
        pz += b[5] * b[6]
    bodies[0][3] = -px / SOLAR_MASS
    bodies[0][4] = -py / SOLAR_MASS
    bodies[0][5] = -pz / SOLAR_MASS

    e1 = energy(bodies)
    for _ in range(n):
        advance(bodies, 0.01)
    e2 = energy(bodies)
    return e1, e2

import time
start = time.time()
e1, e2 = main(10000)
end = time.time()
print(f"Result: {e1:.9f}_{e2:.9f}")
print(f"Time: {int((end - start)*1000)}ms")
